src/Storages/ObjectStorage/DataLakes/DeltaLake/ExpressionVisitor.cpp

```diff
@@ -0,0 +1,750 @@
+
+        if (auto e = data.getException())
+            std::rethrow_exception(e);
+    }
+
+    static void visit(ffi::SharedExpression * expression, ExpressionVisitorData & data)
+    {
+        auto visitor = createVisitor(data);
+        [[maybe_unused]] uintptr_t result = ffi::visit_expression(&expression, &visitor);
+        chassert(result == 0, "Unexpected result: " + DB::toString(result));
+
+        if (auto e = data.getException())
+            std::rethrow_exception(e);
+    }
+
+private:
+    /// At this moment ExpressionVisitor is used only for partition columns,
+    /// where only identifier expressions are allowed (only PARTITION BY col_name, ...),
+    /// therefore we leave several visitor function as not implemented
+    /// (see throwNotImplemented in createVisitor() below).
+    /// They will be implemented once we start using statistics feature from delta-kernel.
+    enum NotImplementedMethod
+    {
+        LT,
+        //LE,
+        GT,
+        //GE,
+        EQ,
+        //NE,
+        DISTINCT,
+        IN,
+        //NOT_IN,
+        ADD,
+        MINUS,
+        MULTIPLY,
+        DIVIDE,
+    };
+    static ffi::EngineExpressionVisitor createVisitor(ExpressionVisitorData & data)
+    {
+        ffi::EngineExpressionVisitor visitor;
+        visitor.data = &data;
+        visitor.make_field_list = &makeFieldList;
+
+        visitor.visit_literal_bool = &visitSimpleLiteral<bool, DB::DataTypeUInt8>;
+        visitor.visit_literal_byte = &visitSimpleLiteral<int8_t, DB::DataTypeInt8>;
+        visitor.visit_literal_short = &visitSimpleLiteral<int16_t, DB::DataTypeInt16>;
+        visitor.visit_literal_int = &visitSimpleLiteral<int32_t, DB::DataTypeInt32>;
+        visitor.visit_literal_long = &visitSimpleLiteral<int64_t, DB::DataTypeInt64>;
+        visitor.visit_literal_float = &visitSimpleLiteral<float, DB::DataTypeFloat32>;
+        visitor.visit_literal_double = &visitSimpleLiteral<double, DB::DataTypeFloat64>;
+
+        visitor.visit_literal_string = &visitStringLiteral;
+        visitor.visit_literal_decimal = &visitDecimalLiteral;
+
+        visitor.visit_literal_timestamp = &visitTimestampLiteral;
+        visitor.visit_literal_timestamp_ntz = &visitTimestampNtzLiteral;
+        visitor.visit_literal_date = &visitDateLiteral;
+        visitor.visit_literal_binary = &visitBinaryLiteral;
+        visitor.visit_literal_null = &visitNullLiteral;
+        visitor.visit_literal_array = &visitArrayLiteral;
+        visitor.visit_literal_struct = &visitStructLiteral;
+        visitor.visit_literal_map = &visitMapLiteral;
+
+        visitor.visit_column = &visitColumnExpression;
+        visitor.visit_struct_expr = &visitStructExpression;
+
+        visitor.visit_or = &visitFunction<DB::FunctionOr>;
+        visitor.visit_and = &visitFunction<DB::FunctionAnd>;
+        visitor.visit_not = &visitFunction<DB::FunctionNot>;
+
+        visitor.visit_is_null = &visitFunction<DB::FunctionIsNull>;
```

Comment:
Better to use FunctionFactory, but okay.

Reply:
And it is used: https://github.com/ClickHouse/ClickHouse/blob/17d95c90f07fdc18b5a356e7475d5acbc4b3e1cd/src/Storages/ObjectStorage/DataLakes/DeltaLake/ExpressionVisitor.cpp#L467